sycl/doc/extensions/proposed/sycl_ext_oneapi_usm_malloc_properties.asciidoc

```diff
@@ -0,0 +1,1468 @@
+auto APtr4 = annotated_malloc_device<int>(N, q, properties{});
+
+// APtr5 is of type annotated_ptr<int*, decltype(properties{usm_kind<sycl::usm::alloc::host>})>
+auto APtr5 = annotated_malloc_host<int>(N, q, properties{});
+
+// APtr6 is of type annotated_ptr<int*, decltype(properties{usm_kind<sycl::usm::alloc::shared>})>
+auto APtr6 = annotated_malloc_shared<int>(N, q, properties{});
+
+static_assert(!std::is_same_v<decltype(APtr4), decltype(APtr5)>);
+static_assert(!std::is_same_v<decltype(APtr4), decltype(APtr6)>);
+static_assert(!std::is_same_v<decltype(APtr5), decltype(APtr6)>);
+----
+
+This extension also introduces USM memory allocation functions with `properties` support that allow the USM memory allocation kind to be specified at runtime.
+In this case, the returned `annotated_ptr` will not have the `sycl::ext::oneapi::experimental::usm_kind` property (unless that property is also passed in).
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+
+properties P4{bar, foo{1}};
+
+// APtr7 is of type annotated_ptr<int*, decltype(properties{bar})>
+auto APtr7 = annotated_malloc<int>(N, q, sycl::usm::alloc::device, P4);
+
+----
+
+If the USM memory allocation kind specified by a parameter to the allocation function is different than the USM memory allocation kind specified by the `sycl::ext::oneapi::experimental::usm_kind` property, the USM memory allocation kind specified by the property takes precedence.
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+
+properties P6{usm_kind<sycl::usm::alloc::device>};
+
+// The memory allocated will be of kind sycl::usm::alloc::device
+auto APtr10 = annotated_malloc<int>(N, q, sycl::usm::alloc::host, P6);
+
+----
+
+
+If the `sycl::ext::oneapi::experimental::usm_kind` property specifies a different USM memory allocation kind than the function supports, the compiler will issue a diagnostic error.
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+
+properties P7{usm_kind<sycl::usm::alloc::device>};
+
+auto APtr11 = annotated_malloc_host<int>(N, q, P7); // Error
+----
+
+The following example uses the compile-time-constant property `alignment`, defined in the link:../proposed/sycl_ext_oneapi_annotated_ptr.asciidoc[sycl_ext_oneapi_annotated_ptr] extension.
+When `alignment` is passed to a USM memory allocation function with `properties` support, it will appear on the returned `annotated_ptr` since it is a compile-time constant property.
+It also informs the runtime to allocate the memory with this alignment.
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
+
+properties P8{alignment<512>};
+// APtr12 is of type annotated_ptr<int, decltype(properties{alignment<512>, usm_kind<sycl::usm::alloc::device>})>
+// The raw pointer of APtr12 is aligned to 512 bytes
+auto APtr12 = annotated_malloc_device<int>(N, q, P8);
+
+properties P9{alignment<1>};
+// Allocate N integers.
+// Alignment must be at least sizeof(int) bytes.
```

Comment:
Does this actually convey what you want to say?  I worry that an alignment of 7 satisfies this requirement as written.  Can we simply say that the resulting pointer has an alignment of sizeof(int) bytes?  Isn't it already part of the definition of alignment that alignment of N can be satisfied by a pointer whose value is any multiple of N (and can therefore be a multiple of XN as well where X is a natural number)?

Reply:
"alignment must be at least sizeof(int) bytes" is a necessary condition but not sufficient.
I can change this to "alignment must be a positive non-zero integer multiple of sizeof(int)".